### ISSUE TITLE
Don't hardcode project, and dcp paths in the assemblies

### ISSUE THREAD

**Issue by radical** [dotnet/aspire#2036]
Currently we add various resolved `dcp` paths, and some projects in `AppHost.dll`. This makes the build output tied to the same machine or at least one with the same layout.

Instead, we should consider adding all this information to a new json (or a custom section in `runtimeconfig.json`) file which gets deployed with the build output. The project paths can be keyed by their names. `launchsettings` needed from the various projects can also be directly referenced or deployed with the AppHost output.

cc @eerhardt @joperezr

**Comment by eerhardt** [comment]
FYI @davidfowl @mitchdenny @danegsta 

I think this would be a good idea. We hit this trying to run tests on a machine that is different than the build machine. If we instead wrote this information out into a json file, the tests could patch up the hard-coded paths and the AppHost would run on the new machine.

We could still do the same `IProjectMetadata` implementation we have today. But instead of returning the hard-coded path:

https://github.com/dotnet/aspire/blob/0d7749b3c34755cc1006ccfee9aa24f7343e4aea/src/Aspire.Hosting/build/Aspire.Hosting.targets#L40-L42

this code instead would say:

```C#
public class ]]>%(ClassName)<![CDATA[ : global::Aspire.Hosting.IProjectMetadata
{
  public string ProjectPath => 
    // load Aspire.AppHost.json file
    // return json.Projects["%(ClassName)"];
```

Similarly we can write the other paths we are hard-coding today into this json file:

```C#
[assembly: AssemblyMetadata("dcpclipath", "C:\\Users\\eerhardt\\.nuget\\packages\\microsoft.developercontrolplane.windows-amd64\\0.1.49\\tools\\dcp.exe")]
[assembly: AssemblyMetadata("dcpextensionpaths", "C:\\Users\\eerhardt\\.nuget\\packages\\microsoft.developercontrolplane.windows-amd64\\0.1.49\\tools\\ext\\")]
[assembly: AssemblyMetadata("dcpbinpath", "C:\\Users\\eerhardt\\.nuget\\packages\\microsoft.developercontrolplane.windows-amd64\\0.1.49\\tools\\ext\\bin\\")]
[assembly: AssemblyMetadata("apphostprojectpath", "C:\\git\\aspire\\playground\\mongo\\Mongo.AppHost")]
[assembly: AssemblyMetadata("aspiredashboardpath", "C:\\git\\aspire\\artifacts\\bin\\Aspire.Dashboard\\Debug\\net8.0\\Aspire.Dashboard.exe")]
```

**Comment by danegsta** [comment]
Seems like this would be a good improvement; the assembly attribute approach was mainly a quick and easy way to make build time knowledge (the location of the DCP and dashboard binaries in the workload) available to the assemblies at runtime, but using a config file would be more flexible.

**Comment by davidfowl** [comment]
Yes, we can do this.

**Comment by joperezr** [comment]
Agreed that this would be great to have, assuming we'd want to get it by preview 4. We should, however, make sure we threat model the feature since IIUC it would mean that someone that could edit this json file could alter which executables are launched by DCP.

**Comment by davidfowl** [comment]
Who is going to work on this?

**Comment by eerhardt** [comment]
I don’t think we’d have to have this in preview4. It’s just internal refactoring that shouldn’t affect users at all.

**Comment by davidfowl** [comment]
Does/will it affect our ability to run tests?

**Comment by mitchdenny** [comment]
Project paths should probably just be relative to the AppHost directory.

**Comment by eerhardt** [comment]
> Does/will it affect our ability to run tests?

This isn't a blocker for running tests. Our plan is to do the build and run of the AppHost on the same machine. Using this approach will also get us coverage of the workload and nuget packages.

The rough idea:

* On the build machine:
    - Build the product, workload, nugets, and tests
    - package all the above up along with the Test Projects
* On the test machine:
    - unpackage everything: workload, nuget packages, test projects
    - Run the tests which will:
        - `dotnet run` the test projects, and then verify things work correctly

**Comment by davidfowl** [comment]
Are we doing this in preview 5?

**Comment by radical** [comment]
```csharp
[assembly: AssemblyMetadata("dcpclipath", "C:\\Users\\eerhardt\\.nuget\\packages\\microsoft.developercontrolplane.windows-amd64\\0.1.49\\tools\\dcp.exe")]
[assembly: AssemblyMetadata("dcpextensionpaths", "C:\\Users\\eerhardt\\.nuget\\packages\\microsoft.developercontrolplane.windows-amd64\\0.1.49\\tools\\ext\\")]
[assembly: AssemblyMetadata("dcpbinpath", "C:\\Users\\eerhardt\\.nuget\\packages\\microsoft.developercontrolplane.windows-amd64\\0.1.49\\tools\\ext\\bin\\")]
```

Will `dcp` be added to the workload at some point? If it isn't in the workload then the tests need to install it to be able to patch the paths for the app to run.

```csharp
[assembly: AssemblyMetadata("aspiredashboardpath", "C:\\git\\aspire\\artifacts\\bin\\Aspire.Dashboard\\Debug\\net8.0\\Aspire.Dashboard.exe")]
```

The dashboard path can be constructed from the installed workload. For in-repo runs, it should use the dashboard from `aritfacts` of course.


```csharp
[assembly: AssemblyMetadata("apphostprojectpath", "C:\\git\\aspire\\playground\\mongo\\Mongo.AppHost")]
```

Why is this path needed? I'm thinking of the case where you build your project and xcopy the output to another machine to run. Does it need access to the source, or some specific files (`launchsettings.json`?) ?

**Comment by radical** [comment]
> Will dcp be added to the workload at some point? If it isn't in the workload then the tests need to install it to be able to patch the paths for the app to run.

Actually, *just* for the tests we can send the dependencies to helix, and patch in the relevant paths.

**Comment by danegsta** [comment]
DCP is in the workload, which is why we added the attributes. We need to know the path to the assemblies at runtime, but we only know the actual path at build time via the workload targets, hence we write attributes at build time to store the data. 

What you’re seeing with the nuget package paths is only for the Aspire repo; we don’t want to require the workload when building or running the tests or playground apps, so we do some logic to pull down a DCP nuget for the appropriate local architecture and version and point to that instead of the workload. 

Now that I think about it, we should be able to pass `—dcp-cli-path` when running the AppHost to load DCP from a known path at runtime, as I did set up some config processing for these properties.

**Comment by joperezr** [comment]
Given this is test-related, I'm tentatively putting it on @radical, but this is currently consider a nice-to-have but not required for P5.

**Comment by davidfowl** [comment]
Moving to the backlog.

**Comment by radical** [comment]
Stale. We are not doing this.
